Möller–Trumbore intersection algorithm

Perhaps the [Möller–Trumbore intersection algorithm](https://en.wikipedia.org/wiki/M%C3%B6ller%E2%80%93Trumbore_intersection_algorithm) would make a good addition to this library? I would be willing to try and implement it; I already made a version of it for a personal project that just uses C and cglm, so it wouldn't be too hard.

I'm not sure if you want a contribution like this though, and I also don't know whether it would be compatible with the license. The Wikipedia page is CC-SA 3.0 (I don't know if in this case a simple notice in a comment would suffice to point out the source, [this is what Wikipedia seems to imply though](https://en.wikipedia.org/wiki/Wikipedia:Reusing_Wikipedia_content#Re-use_of_text_under_Creative_Commons_Attribution-ShareAlike) as long as it would make the result just a modified version and not the entire library by implication a derivative work) but the algorithm originally comes from [this paper](http://www.graphics.cornell.edu/pubs/1997/MT97.pdf) I believe, and I would not know the implications of basing code on a snippet in a paper for the required license.

@recp I've started implementing it, but am running into an issue. I think the inline version works now, but when trying to use the pre-compiled version I get `main.c:(.text+0x25f0): undefined reference to `glmc_line_triangle_intersect'` in my test project.

I'm new to building libraries in C so I'm thinking you might see something obvious that I'm missing. You can look at what I have so far [here](https://github.com/Uwila/cglm/commit/486e3826b1ac24297a73ce62c6c61f3ff207e1a8). If not, I'll probably figure it out eventually.

Aside from that I have a question about taste: The algorithm computes whether there's a collision, and what the distance to the point of collision is. Right now it returns `-1.0f` when there's no collision. I could make it so that instead of returning the distance, it just returns a boolean, and to get the distance you have to pass a pointer to a float for the function to save the distance in. What would have your preference?

Hello @Uwila 

Thanks for starting work on this 🤗

---

> main.c:(.text+0x25f0): undefined reference to `glmc_line_triangle_intersect' in my test project.

Did you link **cglm** to your test project? Which platform do you test on? For instance in **clang**, **gcc** family you add `-lcglm` flag to specify that you are linking/adding a library to your project.`-l`(L) means library, and `-I` (i) means include.

For windows must must see Linker options in Visual Studio. Also it seems you did not add `ray.c` to Visual Studio project, so this file is not compiled on windows...

---

in **glm** there are a few more functions, I don't know what one does what for now:

https://glm.g-truc.net/0.9.4/api/a00179.html
[Github -> glm/gtx/intersect.hpp](https://github.com/g-truc/glm/blob/master/glm/gtx/intersect.hpp)
[Github -> glm/gtx/intersect.inl](https://github.com/g-truc/glm/blob/ea678faff9340ae4a79f50f2edd947141405e128/glm/gtx/intersect.inl)

```C++
bool intersectLineSphere(); /* is this related to ray / ray tracing? */
bool intersectLineTriangle(); /* is this related to ray / ray tracing? */
bool intersectRayPlane();
bool intersectRaySphere();
bool intersectRayTriangle();
```

`glm_line_triangle_intersect()` name seems very long to me. 

For instance in **cglm** all of these are intersection functions:

```C
bool glm_aabb_aabb(vec3 a[2], vec3 b[2]);  /* check if two AABB intersects */
bool glm_aabb_sphere(vec3 box[2], vec4 s); /* check if AABB intersects with sphere */
bool glm_aabb_point(vec3 box[2], vec3 p);  /* check if point is inside of AABB */
bool glm_sphere_sphere(vec4 s1, vec4 s2);  /* check if two sphere intersects */
```

I'm not sure users like it or not.

May I ask what is the difference between `intersectLineTriangle` and `intersectRayTriangle`, if do you know?

Maybe we kan group **ray** related function inside `glm_ray_`, what do you think?

```C
bool glm_ray_intersect_plane();
bool glm_ray_intersect_sphere();
bool glm_ray_intersect_triangle();
```

```C
/* intersect is intended */
bool glm_ray_plane();
bool glm_ray_sphere();
bool glm_ray_triangle();
```

---

> I could make it so that instead of returning the distance, it just returns a boolean, and to get the distance you have to pass a pointer to a float for the function to save the distance in. What would have your preference?

I like this idea. 👍

I want to implement Ray Tracing in my render engine in the future (https://github.com/recp/gk) these features may help me a lot :)

--- a/CREDITS
+++ b/CREDITS
@@ -65,3 +65,9 @@ https://forums.khronos.org/showthread.php/10651-Animation-TCB-Spline-Interpolati
 12. vec2 cross product
 http://allenchou.net/2013/07/cross-product-of-2d-vectors/
 
+13. Ray triangle intersect
+Möller–Trumbore ray-triangle intersection algorithm, from "Fast, Minimum Storage Ray/Triangle Intersection"
+Authors:
+  Thomas Möller (tompa@clarus.se)
+  Ben Trumbore (wbt@graphics.cornell.edu)
+Link to paper: http://webserver2.tecgraf.puc-rio.br/~mgattass/cg/trbRR/Fast%20MinimumStorage%20RayTriangle%20Intersection.pdf

--- a/Makefile.am
+++ b/Makefile.am
@@ -63,7 +63,8 @@ cglm_HEADERS = include/cglm/version.h \
                include/cglm/ease.h \
                include/cglm/curve.h \
                include/cglm/bezier.h \
-               include/cglm/applesimd.h
+               include/cglm/applesimd.h \
+               include/cglm/ray.h
 
 cglm_calldir=$(includedir)/cglm/call
 cglm_call_HEADERS = include/cglm/call/mat4.h \
@@ -84,7 +85,8 @@ cglm_call_HEADERS = include/cglm/call/mat4.h \
                     include/cglm/call/sphere.h \
                     include/cglm/call/ease.h \
                     include/cglm/call/curve.h \
-                    include/cglm/call/bezier.h
+                    include/cglm/call/bezier.h \
+                    include/cglm/call/ray.h
 
 cglm_simddir=$(includedir)/cglm/simd
 cglm_simd_HEADERS = include/cglm/simd/intrin.h \
@@ -147,7 +149,8 @@ libcglm_la_SOURCES=\
     src/sphere.c \
     src/ease.c \
     src/curve.c \
-    src/bezier.c
+    src/bezier.c \
+    src/ray.c
 
 test_tests_SOURCES=\
     test/runner.c \

--- a/docs/source/api.rst
+++ b/docs/source/api.rst
@@ -52,3 +52,4 @@ Follow the :doc:`build` documentation for this
    curve
    bezier
    version
+   ray

--- a/docs/source/ray.rst
+++ b/docs/source/ray.rst
@@ -1,0 +1,31 @@
+.. default-domain:: C
+
+ray
+====
+
+Header: cglm/ray.h
+
+This is for collision-checks used by ray-tracers and the like.
+
+Table of contents (click to go):
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Functions:
+
+1. :c:func:`glm_ray_triangle`
+
+Functions documentation
+~~~~~~~~~~~~~~~~~~~~~~~
+
+.. c:function:: bool  glm_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d)
+
+    Möller–Trumbore ray-triangle intersection algorithm
+
+    Parameters:
+      | *[in]*       **origin**        origin of ray
+      | *[in]*       **direction**     direction of ray
+      | *[in]*       **v0**            first vertex of triangle
+      | *[in]*       **v1**            second vertex of triangle
+      | *[in]*       **v2**            third vertex of triangle
+      | *[in, out]*  **d**             float pointer to save distance to intersection
+      | *[out]*      **intersection**  whether there is intersection

--- a/include/cglm/call.h
+++ b/include/cglm/call.h
@@ -31,6 +31,7 @@ extern "C" {
 #include "call/ease.h"
 #include "call/curve.h"
 #include "call/bezier.h"
+#include "call/ray.h"
 
 #ifdef __cplusplus
 }

--- a/include/cglm/call/ray.h
+++ b/include/cglm/call/ray.h
@@ -14,7 +14,12 @@ extern "C" {
 
 CGLM_EXPORT
 bool
-glmc_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d);
+glmc_ray_triangle(vec3   origin,
+                  vec3   direction,
+                  vec3   v0,
+                  vec3   v1,
+                  vec3   v2,
+                  float *d);
     
 #ifdef __cplusplus
 }

--- a/include/cglm/call/ray.h
+++ b/include/cglm/call/ray.h
@@ -1,0 +1,22 @@
+/*
+ * Copyright (c), Recep Aslantas.
+ *
+ * MIT License (MIT), http://opensource.org/licenses/MIT
+ * Full license can be found in the LICENSE file
+ */
+
+#ifndef cglmc_ray_h
+#define cglmc_ray_h
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "../cglm.h"
+
+CGLM_EXPORT
+bool
+glmc_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d);
+    
+#ifdef __cplusplus
+}
+#endif
+#endif /* cglmc_ray_h */

--- a/include/cglm/cglm.h
+++ b/include/cglm/cglm.h
@@ -30,5 +30,6 @@
 #include "ease.h"
 #include "curve.h"
 #include "bezier.h"
+#include "ray.h"
 
 #endif /* cglm_h */

--- a/include/cglm/ray.h
+++ b/include/cglm/ray.h
@@ -42,7 +42,7 @@ glm_ray_triangle(vec3   origin,
                  float *d) {
   vec3        edge1, edge2, p, t, q;
   float       det, inv_det, u, v, dist;
-  const float epsilon = 0.000001;
+  const float epsilon = 0.000001f;
 
   glm_vec3_sub(v1, v0, edge1);
   glm_vec3_sub(v2, v0, edge2);
@@ -54,32 +54,28 @@ glm_ray_triangle(vec3   origin,
   if (det > -epsilon && det < epsilon)
     return 0;
 
-  inv_det = 1.0 / det;
+  inv_det = 1.0f / det;
 
   glm_vec3_sub(origin, v0, t);
 
   u = inv_det * glm_vec3_dot(t, p);
 
-  if (u < 0.0 || u > 1.0)
+  if (u < 0.0f || u > 1.0f)
     return 0;
 
   glm_vec3_cross(t, edge1, q);
 
   v = inv_det * glm_vec3_dot(direction, q);
 
-  if (v < 0.0 || u + v > 1.0)
+  if (v < 0.0f || u + v > 1.0f)
     return 0;
 
   dist = inv_det * glm_vec3_dot(edge2, q);
-  if (dist > epsilon) {
-    if (d != NULL) {
+
+  if (d)
       *d = dist;
-    }
 
-    return 1;
-  }
-
-  return 0;
+  return dist > epsilon;
 }
 
 #endif

--- a/include/cglm/ray.h
+++ b/include/cglm/ray.h
@@ -7,7 +7,12 @@
 
 /*
  Functions:
-   CGLM_INLINE bool glm_line_triangle_intersect(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d);
+   CGLM_INLINE bool glm_line_triangle_intersect(vec3   origin,
+                                                vec3   direction,
+                                                vec3   v0,
+                                                vec3   v1,
+                                                vec3   v2,
+                                                float *d);
 */
 
 #ifndef cglm_ray_h
@@ -29,7 +34,12 @@
 
 CGLM_INLINE
 bool
-glm_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d) {
+glm_ray_triangle(vec3   origin,
+                 vec3   direction,
+                 vec3   v0,
+                 vec3   v1,
+                 vec3   v2,
+                 float *d) {
   vec3        edge1, edge2, p, t, q;
   float       det, inv_det, u, v, dist;
   const float epsilon = 0.000001;

--- a/include/cglm/ray.h
+++ b/include/cglm/ray.h
@@ -30,37 +30,46 @@
 CGLM_INLINE
 bool
 glm_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d) {
-    const float epsilon = 0.000001;
-    vec3 edge1, edge2, p, t, q;
-    float det, inv_det, u, v;
+  vec3        edge1, edge2, p, t, q;
+  float       det, inv_det, u, v, dist;
+  const float epsilon = 0.000001;
 
-    glm_vec3_sub(v1, v0, edge1);
-    glm_vec3_sub(v2, v0, edge2);
+  glm_vec3_sub(v1, v0, edge1);
+  glm_vec3_sub(v2, v0, edge2);
 
-    glm_vec3_cross(direction, edge2, p);
+  glm_vec3_cross(direction, edge2, p);
 
-    det = glm_vec3_dot(edge1, p);
+  det = glm_vec3_dot(edge1, p);
 
-    if (det > -epsilon && det < epsilon)
-        return 0;
-    inv_det = 1.0 / det;
+  if (det > -epsilon && det < epsilon)
+    return 0;
 
-    glm_vec3_sub(origin, v0, t);
+  inv_det = 1.0 / det;
 
-    u = inv_det * glm_vec3_dot(t, p);
+  glm_vec3_sub(origin, v0, t);
 
-    if (u < 0.0 || u > 1.0)
-        return 0;
+  u = inv_det * glm_vec3_dot(t, p);
 
-    glm_vec3_cross(t, edge1, q);
+  if (u < 0.0 || u > 1.0)
+    return 0;
 
-    v = inv_det * glm_vec3_dot(direction, q);
+  glm_vec3_cross(t, edge1, q);
 
-    if (v < 0.0 || u + v > 1.0)
-        return 0;
+  v = inv_det * glm_vec3_dot(direction, q);
 
-    *d = inv_det * glm_vec3_dot(edge2, q);
-    return *d > epsilon;
+  if (v < 0.0 || u + v > 1.0)
+    return 0;
+
+  dist = inv_det * glm_vec3_dot(edge2, q);
+  if (dist > epsilon) {
+    if (d != NULL) {
+      *d = dist;
+    }
+
+    return 1;
+  }
+
+  return 0;
 }
 
 #endif

--- a/include/cglm/ray.h
+++ b/include/cglm/ray.h
@@ -1,0 +1,66 @@
+/*
+ * Copyright (c), Recep Aslantas.
+ *
+ * MIT License (MIT), http://opensource.org/licenses/MIT
+ * Full license can be found in the LICENSE file
+ */
+
+/*
+ Functions:
+   CGLM_INLINE bool glm_line_triangle_intersect(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d);
+*/
+
+#ifndef cglm_ray_h
+#define cglm_ray_h
+
+#include "vec3.h"
+
+/*!
+ * @brief Möller–Trumbore ray-triangle intersection algorithm
+ * 
+ * @param[in] origin         origin of ray
+ * @param[in] direction      direction of ray
+ * @param[in] v0             first vertex of triangle
+ * @param[in] v1             second vertex of triangle
+ * @param[in] v2             third vertex of triangle
+ * @param[in, out] d         distance to intersection
+ * @param[out] intersection  whether there is intersection
+ */
+
+CGLM_INLINE
+bool
+glm_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d) {
+    const float epsilon = 0.000001;
+    vec3 edge1, edge2, p, t, q;
+    float det, inv_det, u, v;
+
+    glm_vec3_sub(v1, v0, edge1);
+    glm_vec3_sub(v2, v0, edge2);
+
+    glm_vec3_cross(direction, edge2, p);
+
+    det = glm_vec3_dot(edge1, p);
+
+    if (det > -epsilon && det < epsilon)
+        return 0;
+    inv_det = 1.0 / det;
+
+    glm_vec3_sub(origin, v0, t);
+
+    u = inv_det * glm_vec3_dot(t, p);
+
+    if (u < 0.0 || u > 1.0)
+        return 0;
+
+    glm_vec3_cross(t, edge1, q);
+
+    v = inv_det * glm_vec3_dot(direction, q);
+
+    if (v < 0.0 || u + v > 1.0)
+        return 0;
+
+    *d = inv_det * glm_vec3_dot(edge2, q);
+    return *d > epsilon;
+}
+
+#endif

--- a/src/ray.c
+++ b/src/ray.c
@@ -1,0 +1,8 @@
+#include "../include/cglm/cglm.h"
+#include "../include/cglm/call.h"
+
+CGLM_EXPORT
+bool
+glmc_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d) {
+    return glm_ray_triangle(origin, direction, v0, v1, v2, d);
+}

--- a/src/ray.c
+++ b/src/ray.c
@@ -3,6 +3,11 @@
 
 CGLM_EXPORT
 bool
-glmc_ray_triangle(vec3 origin, vec3 direction, vec3 v0, vec3 v1, vec3 v2, float *d) {
+glmc_ray_triangle(vec3   origin,
+                  vec3   direction,
+                  vec3   v0,
+                  vec3   v1,
+                  vec3   v2,
+                  float *d) {
     return glm_ray_triangle(origin, direction, v0, v1, v2, d);
 }

--- a/win/cglm.vcxproj
+++ b/win/cglm.vcxproj
@@ -34,6 +34,7 @@
     <ClCompile Include="..\src\plane.c" />
     <ClCompile Include="..\src\project.c" />
     <ClCompile Include="..\src\quat.c" />
+    <ClCompile Include="..\src\ray.c" />
     <ClCompile Include="..\src\sphere.c" />
     <ClCompile Include="..\src\vec2.c" />
     <ClCompile Include="..\src\vec3.c" />
@@ -61,6 +62,7 @@
     <ClInclude Include="..\include\cglm\call\plane.h" />
     <ClInclude Include="..\include\cglm\call\project.h" />
     <ClInclude Include="..\include\cglm\call\quat.h" />
+    <ClInclude Include="..\include\cglm\call\ray.h" />
     <ClInclude Include="..\include\cglm\call\sphere.h" />
     <ClInclude Include="..\include\cglm\call\vec2.h" />
     <ClInclude Include="..\include\cglm\call\vec3.h" />
@@ -80,6 +82,7 @@
     <ClInclude Include="..\include\cglm\plane.h" />
     <ClInclude Include="..\include\cglm\project.h" />
     <ClInclude Include="..\include\cglm\quat.h" />
+    <ClInclude Include="..\include\cglm\ray.h" />
     <ClInclude Include="..\include\cglm\simd\arm.h" />
     <ClInclude Include="..\include\cglm\simd\avx\affine.h" />
     <ClInclude Include="..\include\cglm\simd\avx\mat4.h" />

--- a/win/cglm.vcxproj.filters
+++ b/win/cglm.vcxproj.filters
@@ -92,6 +92,9 @@
     <ClCompile Include="..\src\vec2.c">
       <Filter>src</Filter>
     </ClCompile>
+    <ClCompile Include="..\src\ray.c">
+      <Filter>src</Filter>
+    </ClCompile>
   </ItemGroup>
   <ItemGroup>
     <ClInclude Include="..\src\config.h">
@@ -122,6 +125,9 @@
       <Filter>include\cglm\call</Filter>
     </ClInclude>
     <ClInclude Include="..\include\cglm\call\vec4.h">
+      <Filter>include\cglm\call</Filter>
+    </ClInclude>
+    <ClInclude Include="..\include\cglm\call\ray.h">
       <Filter>include\cglm\call</Filter>
     </ClInclude>
     <ClInclude Include="..\include\cglm\simd\avx\affine.h">
@@ -239,6 +245,9 @@
       <Filter>include\cglm\call</Filter>
     </ClInclude>
     <ClInclude Include="..\include\cglm\ease.h">
+      <Filter>include\cglm</Filter>
+    </ClInclude>
+    <ClInclude Include="..\include\cglm\ray.h">
       <Filter>include\cglm</Filter>
     </ClInclude>
     <ClInclude Include="..\include\cglm\simd\arm.h">

